[1단계 - 자동차 경주 구현] 메리(최승원) 미션 제출합니다.

swonny opened the pull request:
부족하지만 잘 부탁드리겠습니다!
감사합니다.

swonny committed:
test: 문자열 분리 테스트

swonny committed:
test: 문자열 삭제 테스트

swonny committed:
test: 특정 인덱스 문자 추출 테스트

swonny committed:
test: 스트링 인덱스 범위 오류 테스트

swonny committed:
test: DisplayName 어노테이션 추가

swonny committed:
test: HashSet 사이즈 테스트

swonny committed:
test: 매개변수 테스트

swonny committed:
test: 매개변수 테스트

swonny committed:
docs: 구현 기능 목록 작성

swonny committed:
feat: 자동차 이름 입력

- 자동차 이름을 입력 받는다.
- 자동차를 생성한다.

swonny committed:
docs: 기능 목록 구현에서 자동차 이름 입력받기 체크

swonny committed:
feat: 시도 횟수 입력 기능 구현

swonny committed:
docs: 구현 기능 목록 체크

swonny committed:
feat: 자동차 이동 기능 구현

swonny committed:
refactor: 자동차 이동 로직 리팩토링

swonny committed:
docs: 기능 구현 목록에서 자동차 이동 기능 체크

swonny committed:
feat: 자동차 위치 출력 기능 구현

swonny committed:
docs: 기능 구현 목록에서 자동차 위치 출력 기능 체크

swonny committed:
fix: 포지션 시작 지점 변경, 출력 포맷 맞추기

swonny committed:
docs: 구현 기능 목록 체크

swonny committed:
feat: 우승자 판별 구현

swonny committed:
docs: 구현 기능 목록 체크 및 수정

swonny committed:
feat: 자동차 이름 5글자 제한 예외 처리 기능 구현

swonny committed:
docs: 자동차 이름 5글자 제한 기능 구현 체크

swonny committed:
feat: 이동횟수가 숫자가 아닌 경우 예외 처리

swonny committed:
docs: 이동 횟수 예외 처리 구현 체크

swonny committed:
refactor: 자동차 이름 예외처리 기능 메서드 분리

swonny committed:
docs: 이동 횟수 입력 예외처리 추가

swonny committed:
feat: 이동횟수 입력 0 이하인 경우 예외 처리

swonny committed:
docs: 이동 횟수 입력 받기 예외 처리 구현 체크

swonny committed:
docs: 기능 구현 목록에 자동차 이름 예외 케이스 추가

swonny committed:
feat: 자동차 이름 입력 시, 쉼표 2개 이상 연속 입력 예외 처리

swonny committed:
docs: 기능 구현 목록에서 자동차 이름 입력 예외 처리 체크

swonny committed:
feat: 자동차 생성자 리팩토링

- 자동차 게임 우승자 테스트를 위한 생성자 추가

swonny committed:
test: 자동차 게임 우승자 구하기 테스트

swonny committed:
docs: 자동차 이름 입력 예외 케이스 추가

swonny committed:
feat: 자동차 한 대 이하로 입력한 경우 예외 처리

- 게임을 진행하기 위해 자동차가 두 대 이상 필요하기 때문에 예외처리를 진행한다.

swonny committed:
docs: 자동차 이름 입력 예외 기능 구현 체크

swonny committed:
test: CarRepository의 findAll 메서드 테스트 추가

- findAll() : 레포지토리에 저장된 모든 자동차 반환하는 메서드

swonny committed:
test: 우승자 판별을 위한 max position 구하기 테스트

swonny committed:
fix: 테스트 시, 레포지토리의 클래스 필드에 자동차가 누적되어 발생하는 오류 해결

swonny committed:
test: 자동차 생성시 이름 길이 예외 처리 테스트

swonny committed:
test: 자동차 개수 예외 테스트

swonny committed:
refactor: 메소드 분리, 불필요한 import 구문 제거, 상수화

include42 commented:
안녕하세요, 메리! 이번에 리뷰를 하게 된 카프카입니다.
리뷰 요청 확인하였습니다. 오늘 저녁 10시~11시 사이에 리뷰 진행 예정입니다.

미션 마무리까지 고생하셨습니다. 
내일 리뷰 확인하시고, 궁금한 점이 있다면 편하게 말씀해주세요. 
이번 리뷰를 통해 함께 성장할 수 있다면 좋겠습니다. 👍

include42 commented on docs/README.md:
```diff
@@ -0,0 +1,35 @@
+# 자동차경주 게임
```
명세가 명확해서 좋았습니다. 번거롭더라도, 명세를 미리 작성해두고 체크하면서 프로젝트를 진행하는게 좋습니다. 💯 

다만 궁금했던 점이, 혹시 docs 패키지 안에 따로 README를 만든 이유가 있었을까요?
최상위 경로에 있는 README에 작성하면 더 편하게 확인할 수 있는만큼 좋지만,
docs라는 패키지에 따로 정리해 두신 것도 명확하고 좋다고 느껴서, 그 이유를 질문합니다.

swonny replied:
최상위 경로의 README에는 프로그램에 대한 개요, 실행 방법, 결과 등이 포함되어야 있어야 한다고 생각했습니다.
기능 명세는 프로그램을 만드는 입장에서만 필요한 정보이기 때문에 docs라는 패키지에 포함시켰는데요,
최상위 경로에 있는 README에 프로그램 개요, 실행 결과 등을 추가하거나 미션을 하는 동안에는 기능 명세를 최상위에 작성해두어도 좋을 것 같다는 생각이 듭니다!!

include42 replied:
이 부분은 정말 잘해주셨네요. 최상위 경로의 README의 목적은 말씀하신 내용이 들어가는 것이 옳다고 저도 생각합니다.
앞으로 다른 페어들과 미션을 하면서 이 부분에 대해 이야기를 더 많이 해보길 권해요. 
저는 메리가 제시한 README의 철학에 공감합니다. 💯

include42 commented on src/main/java/constant/Constants.java:
```diff
@@ -0,0 +1,16 @@
+package constant;
+
+public class Constants {
```
상수를 따로 정리한 부분이 깔끔하게 느껴졌습니다. 
다만 코드를 보다 보니, 여기에 속하지 않는 상수 (예: GameController의 `CAR_NAME_DELIMITER `)도 있더군요.
Constants에 저장한 상수와 그렇지 않은 상수의 기준이 있을까요?

include42 replied:
수정을 요구하는 부분은 아닙니다. 혹시 생각한 상수 관리의 기준이 있는지, 그에 맞게 관리되고 있는지 질문드립니다.

swonny replied:
먼저 Constants에서는 OutputView 클래스에서 사용되는 상수만을 관리하도록 작성했습니다.
OutputView에서 사용되는 문자열을 하드코딩 하지 않고 상수로 관리하고자 했고,
각 메서드에서 사용되는 문자열을 모두 상수로 만들자니 클래스의 필드가 너무 많아진다고 생각해 상수를 분리하게 되었습니다.

그런데 남겨주신 리뷰를 확인하고 페어와 함께 고민해보면서 생각이 조금 바뀌어 Constants는 삭제하였습니다.
재사용의 가능성이 있으며 도메인과 관련된 매직넘버는 상수로 관리하는 것이 좋다는 것에는 저희 둘 다 동의했지만
'OutputView와 같이 일회성으로 사용되는 문자열을 상수로 관리해주어야 할까?'에 대한 의문이 들었기 때문입니다.

또한 GameController의 `CAR_NAME_DELIMITER`의 경우,
InputView에서 전달 받은 자동차 이름을 분리하는 것은 GameController의 책임이기 때문에 GameController가 알고 있어야 한다고 생각해 분리해주지 않았었는데요.
고민해본 결과 자동차 이름을 분리하여 생성하는 것은 Controller의 책임이 아니라고 생각했습니다.
따라서 이 부분은 CarFactory클래스를 생성하여 문자열을 분리하도록 리팩토링하였고,
`CAR_NAME_DELIMITER`는 자동차 객체를 생성하는 CarFactory가 관리하도록 하였습니다.

저와 페어는 도메인과 관련 있는 매직넘버는 상수로 관리하되, View에서 일회성으로 사용되는 문자열은 하드코딩을 하는 것으로 결론을 지었는데, 카프카님의 상수를 관리하는 기준이 있으신지 궁금합니다!

include42 replied:
앞으로 우테코에서 진행할 spring 프로젝트에서는 문자열 상수(예: 에러 메시지, 프로젝트 정보 등)를 저장하는 기법이 보다 복잡하기 때문에, 이에 대해 지금 설명하기보다는 앞으로 프로젝트를 진행하면서 학습해 보는 것이 좋다고 생각합니다.

다만 클래스에서 가지는 매직 넘버와 같은 값은, 저의 경우 해당 클래스 내에서 관리하도록 하고 있어요.
코멘트 적어주신 것처럼, 책임에 해당하는 곳에서 가지는 것이 관리에 있어 유연할 수 있기 때문입니다.

저는 메리가 페어와 논의해 내린 결론에 동의합니다.
물론 뷰에서 사용하는 문자열을 하드코딩하는 것에 대해서는 의견이 좀 갈릴 수 있을 것 같아요.
이 부분 역시, 다음 프로젝트를 진행할 때 새 페어와 논의를 다시 해보길 권할게요.
이렇게 논의하며 고민하는 과정을 거친 것이 매우 좋은 문제해결 방법이라고 생각합니다. 💯

include42 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+    private final RacingCarGame racingCarGame;
+
+    public GameController(OutputView outputView) {
+        this.inputView = new InputView(outputView);
```
코드를 읽어보면서, InputView가 OutputView를 포함하는 구조가 직관적으로 이해되지 않았습니다.
GameController에서 두개를 모두 가지고 있어서 더욱 헷갈리는 측면이 있었네요.

InputView는 단순히 입력을 받아서 반환하는 메소드들로 구성하는게 가능할까요?

swonny replied:
InputView가 OutputView를 포함하도록 한 이유는
입력값을 받을 때 안내문구를 출력하는 것이 InputView의 책임이라고 생각했기 때문입니다.
예를 들어 자동차 이름을 입력받아야 하는 경우 반드시 `자동차 이름을 입력해주세요.`와 같은 안내 문구가 출력되어야 합니다.
따라서 입력 안내문구를 출력하는 것을 한 묶음으로 보고 InputView에서 OutputView를 포함하도록 구현하였습니다.

그런데 이 부분도 남겨주신 리뷰를 보고 고민해본 결과 메서드의 책임이 어디까지인지에 대한 생각은 주관적인 것이기 때문에 클래스 또는 메소드명을 기준으로 책임을 분리하는 것이 가장 직관적이고 객관적으로 책임을 분리하는 방법이라고 생각했습니다.

따라서 InputView와 OutputView는 메소드명에 해당하는 기능만을 수행하고,
관련된 기능을 묶어주는 것은 Controller에서 담당하는 구조로 리팩토링을 진행하였습니다.

include42 replied:
이 부분도 고민이 많았을 것 같아요. 제가 InputView의 구조 개선을 요구했지만, 메리가 말한 대로 한 묶음으로 보는 것도 가능하기 때문입니다. 그럼에도 개선한 코드를 보면, 책임이 명확히 분산되는 면이 있다고 생각합니다. 잘 고쳐주셨습니다. 👍

include42 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+            CarRepository.updateCars(makeCars(carNames));
+        } catch (IllegalArgumentException illegalArgumentException) {
+            outputView.printException(illegalArgumentException.getMessage());
+            addCars(inputView.readCarNames());
```
재귀로 깔끔하게 구현해서 좋았습니다. 💯

include42 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private boolean validateCarName(String carName) {
+        if (carName.isEmpty()) {
```
현재는 차 이름을 공백으로 입력할 경우 실행이 되네요.
이것이 정상 작동이 아니라고 판단된다면, `isBlank`에 대해 검색해보기를 권합니다.

swonny replied:
이 부분도 반영하여 리팩토링 진행하였습니다!

include42 commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,45 @@
+        this.position = position;
+    }
+
+    private void validateName(String name) {
```
앞서 이름에 대한 validate 로직이 controller에 있었는데, 여기에도 validate하는 로직이 나뉘어져 있네요.
controller에서 통합해서 validate를 해도 괜찮지 않을까요? 
혹은 validator라는 유틸 객체를 만들어서 validate 메소드들을 모아도 괜찮고요.

이부분에 대해서는 혹시 의견이 다르다면 편하게 코멘트 남겨주세요. 
나눈 이유가 있다면, 그에 따라 코드를 작성할 수도 있다고 생각합니다.

swonny replied:
같은 CarName에 대한 validate라도 Controller에서는 콤마로 구분된 차 이름 리스트를 valiate한다면(빈 문자열이 포함되어 있는지, 콤마가 연속으로 두 번 입력되었는지 등),
Car 객체의 이름을 저장할 때는 자동차 이름이 5글자 이하인지를 확인하는 책임이 다르다고 생각하여 CarName에 대한 로직을 분리하여 작성하였습니다.

그런데 이 부분도 마찬가지로 주관적인 기준으로 코드가 분산되어 있다면 코드를 직관적으로 이해하기 어려울 것이라고 생각했습니다.
따라서 Validator를 만들어 한 곳에서 validate를 진행하도록 리팩토링 하였습니다.

include42 commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void move(MoveStrategy moveStrategy) {
+        if (moveStrategy.isMovable()) {
+            position++;
```
++를 하는 것도 명확하지만, move할때 추가되는 속도값을 상수로 클래스 내에 정의해도 괜찮지 않을까요?
나중에 명세가 변경되어서 속도값이 변경된다면, 상수를 변경하는 쪽이 관리하기 편할 것으로 생각됩니다.

swonny replied:
<img width="685" alt="image" src="https://user-images.githubusercontent.com/81925468/218234731-78a30fbc-6c16-4309-8903-1344adc8cafb.png">

이 부분은 추가되는 속도값을 상수값을 정의하여 `position += 상수값`을 하도록 변경하는 것이 좋을 것 같다는 말씀으로 이해했는데요,
사진에서 처럼 move 앞 기호가 보이지 않아 우선 코멘를 남깁니다! 제가 이해한 게 맞다면 이 방향으로 수정해보겠습니다.

include42 replied:
코멘트가 깨져서 분명 수정을 했는데, 다시 보니 또 깨져있군요 😓 
이야기해주신 방향이 맞습니다. 수정 내용은 확인해 보겠습니다 👍

include42 commented on src/main/java/racingcar/CarRepository.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static List<Car> findAll() {
+        return Collections.unmodifiableList(cars);
```
unmodifiableList로 반환한 부분 매우 좋은 습관입니다. 💯 💯

include42 commented on src/main/java/racingcar/CarRepository.java:
```diff
@@ -0,0 +1,35 @@
+        return Collections.unmodifiableList(cars);
+    }
+
+    public static List<String> findSamePositionWith(int maxPosition) {
```
인자로 maxPosition을 받는데 메소드명이 findSamePositionWith 라면 이상하지 않을까요?
인자명을 position으로 바꾸거나, 혹은 메소드명을 findMaxPositionCar로 바꾸거나 하는 것이 명확할 것으로 생각됩니다.

swonny replied:
리뷰 남겨주신 메소드명으로 리팩토링을 진행하였습니다.
(메소드명을 고민하는 과정은 정말 어려운 일인 것 같습니다..!🥲)

include42 replied:
메소드명 짓기와 변수명 짓기가 정말 어렵지요 😓  저도 늘 고민하는 지점입니다.
그래도 늘 기획/명세에 맞게, 조금 더 길어지더라도 명확하게, 그리고 새로운 팀원이 와도 바로 이해할 수 있게 작성한다고 생각하며 고민하는 수밖에 없다고 생각합니다. 화이팅입니다 💪

include42 commented on src/main/java/racingcar/MoveStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar;
+
+public interface MoveStrategy {
```
전략 패턴 👍 👍 
해당 패턴을 잘 써주셔서 놀랐습니다.

include42 commented on src/main/java/racingcar/RacingCarGame.java:
```diff
@@ -0,0 +1,25 @@
+
+    public void repeatRounds() {
+        for (Car car : CarRepository.findAll()) {
+            car.move(new RandomBasedStrategy());
```
`randomBasedStrategy`를 반복문이 수행될 때마다 생성해 줄 필요가 있을까요?

swonny replied:
이 부분은 저도 PR을 제출한 이후 재확인하면서 아쉬웠던 부분이었습니다. 변수로 저장하여 사용하는 방식으로 변경해보겠습니다.

include42 commented on src/test/java/racingcar/CarRepositoryTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideCars")
```
벌써 테스트 관련 어노테이션을 잘 소화하고 있어서, 코드가 많이 깔끔해졌네요. 👍 👍

include42 commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,37 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+class CarTest {
```
CarTest 에서 성공/실패 전략을 lambda로 작성한 부분 매우 인상적이었습니다.
테스트 코드를 정말 잘 작성하시네요.

다만 테스트명이 단순해서 바로 파악이 안되는 점이 아쉬웠는데, 
`@DisplayName`을 통해 테스트 수행 목적을 적어주는 게 좋다고 생각합니다. (추가해 본 다음 테스트를 실행해 보세요!)

swonny replied:
감사합니다! `@DisplayName`을 사용하여 작성해보겠습니다.

include42 commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"merrrrrrry", "pooooooobi", "pooooo", ""})
```
`""` 인 경우에 대해 테스트를 작성해 주셨는데, 혹시 null인 경우도 필요하지 않을까요?
`@NullAndEmptySource` 에 대해 검색해서 테스트를 작성해보길 권합니다.

swonny replied:
감사합니다. 해당 어노테이션을 공부해보고 적용해보겠습니다!

include42 reviewed (commented):
안녕하세요, 메리! 리뷰어 카프카입니다.
코드를 깔끔하게 잘 작성해주셨네요. 인상적인 부분들이 많았습니다. 👍 
몇 가지 수정하면 좋은 부분에 코멘트를 남겼습니다. 
확인해보시고, 이해가 안되는 부분은 편하게 코멘트 내지 dm을 부탁드립니다.

include42 reviewed (changes requested):
안녕하세요, 메리! 리뷰어 카프카입니다.
코드를 깔끔하게 잘 작성해주셨네요. 인상적인 부분들이 많았습니다. 👍
몇 가지 수정하면 좋은 부분에 코멘트를 남겼습니다.
확인해보시고, 이해가 안되는 부분은 편하게 코멘트 내지 dm을 부탁드립니다.

swonny committed:
refactor: 상수 클래스 삭제

- 한번만 사용되는 상수는 하드코딩으로 변경:
프로그램이 복잡해지는 것을 방지하기 위해 두 번 이상 사용될 때 상수화 하는 방향으로 진행하려고 합니다.

swonny committed:
refactor: 인풋뷰, 아웃풋뷰의 역할 분리

- 인풋뷰가 아웃풋뷰를 가지고 있어서 로직 파악하는데에 힘들기 때문에, 인풋뷰가 아웃풋뷰에 접근하지 못하도록 변경

swonny committed:
refactor: 자동차 이름에 공백이 포함되어 있는 경우 예외 처리

swonny committed:
refactor: 자동차 이름 예외 처리 클래스 분리 및 각 이름에 공백이 포함될 경우 예외처리

swonny committed:
refactor: 이동할 거리를 나타내는 매직넘버 상수화

swonny committed:
refactor: 최대 포지션과 같은 포지션인 차 이름을 반환하는 메서드 이름 개선

swonny committed:
refactor: 랜덤 값을 구현하는 전략패턴의 구현체를 라운드를 반복하는 메서드 안에서 변수로 따로 분리

- 불필요한 객체 생성을 피하기 위해 변수로 분리합니다.

swonny committed:
refactor: 테스트 코드의 가독성을 위해 @Display 어노테이션 추가

swonny committed:
refactor: 자동차 이름 예외 테스트의 @NullAndEmpty 추가 및 자동차 이름 Null 예외 처리 기능 추가

swonny committed:
refactor: CarFactory 클래스 생성 및 컨트롤러에서 자동차 추가하는 메서드 명 개선

swonny committed:
refactor, test: 시도 횟수 예외 처리 클래스 분리, 빈 문자열 예외 처리 추가, 예외 테스트 추가

swonny committed:
refactor(GameController): startRacing 메소드 내부에 사용되지 않는 인스턴스(round) 제거 및 while문으로 변경

swonny committed:
refactor: 함수형 인터페이스에@FunctionalInterface 어노테이션 추가

swonny committed:
test: 자동차 이름 유효성 검사 기능 테스트 추가

swonny committed:
test: 자동차 이름 중복, 자동차 한 대 이하 입력 유효성 검사 기능 테스트 추가

swonny reviewed (commented):
안녕하세요 카프카님!
우테코 5기 메리입니다🎄

꼼꼼하게 작성해주신 리뷰 잘 보았습니다!
남겨주신 리뷰를 확인하고 페어와 함께 열심히 토론하면서 다시 한번 생각을 정립할 수 있는 기회가 되었습니다.
감사합니다!!

남겨주신 피드백에 대한 고민을 해본 뒤 반영하여 리팩토링을 진행하였습니다.
부족하지만 두 번째 리뷰도 잘 부탁드리겠습니다!

include42 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private void startRacing(int numberOfTry) {
+        while (numberOfTry-- > 0) {
```
`--, ++`을 사용해 코드를 단축하는 것은, 자칫 다른 팀원이 코드를 잘못 이해하도록 할 수 있습니다.
특히나 while 조건문 안에 연산을 넣게 되면, 이를 머리 속으로 구현해보는데에 시간이 많이 소요되지요.

저라면, 코드가 좀 예쁘지 않아 보이더라도 for문을 써서 읽는 사람이 바로 이해하도록 작성할 것 같습니다.

include42 commented on src/main/java/factory/CarFactory.java:
```diff
@@ -0,0 +1,21 @@
+public class CarFactory {
+    private static final String CAR_NAME_DELIMITER = ",";
+
+    private CarFactory() {
```
Factory 패턴에 대해 그사이 학습을 수행했군요! 
이 부분은 저번 리뷰때 권하볼까 고민이 되었던 부분이었는데, 아주 잘 수행해 주었습니다.

디자인 패턴에 대한 학습은 개발을 수행하는데에 도움이 많이 됩니다.
이번에 사용한 전략 패턴과 팩토리 패턴에 대해, 노션 등에 간단히 정리해두면 앞으로 개발부터 취준까지 도움을 많이 받을거에요 👍 👍

include42 commented on src/main/java/validator/CarNameValidator.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Objects;
+
+public class CarNameValidator {
```
validator 작업도 깔끔하게 잘 해주셨습니다.
검증을 잘 수행하고 있는 걸 수행해보면서 확인할 수 있었습니다 👍

include42 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,31 @@
+        try {
+            return BUFFERED_READER.readLine();
+        } catch (IOException ioException) {
+//            outputView.printException(ioException.getMessage());
```
이 부분은 제가 개발하고 있는 팀에서 유지하려고 하는 룰인데요, 
저희 팀의 경우 이전에 쓰던 코드를 삭제할 때에는 주석으로 남기지 않으려고 노력하고 있어요.

git 기록 등을 통해 삭제된 내용은 필요시 확인할 수 있는데, 주석으로 남겨둠으로 인해 읽는 사람에게도 혼동을 주기 때문입니다.
예를 들어, 레거시 코드를 보는데 5년 전에 주석으로 누군가 처리한 한줄의 코드를 본다면 어떨까요? 해당 코드를 이해하는데에 추가적인 시간이 필요해 집니다. (정말 중요해서 주석으로 한건지, 당시에 지우는 걸 잊었는지 판단하기까지 말이죠)

include42 commented on src/test/java/validator/NumberOfTryValidatorTest.java:
```diff
@@ -0,0 +1,37 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+class NumberOfTryValidatorTest {
+
```
전반적으로 테스트를 잘 작성해주셨고, 성공적으로 수행되는 것도 확인하였습니다.
여기에 하나 더 코멘트를 드리자면, 테스트를 작성할때는 성공하는 경우에 대한 테스트도 추가해두는 것이 좋습니다.

만약 해당 테스트와 관련된 클래스를 수정할 경우, 실패 케이스 뿐 아니라 성공 케이스까지 테스트해봐야 문제가 없다는 걸 확신할 수 있거든요. 그리고 성공/실패 케이스들에 대한 테스트를 본다면 코드가 어떻게 작동하는지 이해하기도 더 쉽지요.

include42 reviewed (approved):
안녕하세요 메리, 리뷰어 카프카입니다.
늦은 시간에 리뷰를 드리게 되었네요.

저번에 리뷰하면서 개선사항 요구도 많았고 질문도 많았는데,
이에 대해 고민해면서 리팩토링하시는 모습이 인상적이었습니다. 👍 
이번 미션을 진행하면서 정말 많은 내용을 학습하셨다고 생각되는데,
이를 잘 활용해서 앞으로의 미션도 수월하게 진행해나가시면 좋겠습니다.

이번 미션은 approve 하겠습니다. 2단계에서 다시 뵙겠습니다 😄
